### PR TITLE
Issue 16557 index relationships as nested objects

### PR DESCRIPTION
**Pull request by nollymar**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 1]
```diff
@@ -1389,152 +1389,201 @@ public Identifier getRelatedIdentifier(Contentlet contentlet,String relationship
...
+                        try {
+                            result.add(findContentletByIdentifierAnyLanguage(
+                                    (String) child.get("identifier")));
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                    });
+                }
+            } catch (Exception e) {
+                Logger.error(this, e.getMessage(), e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Please log the content's Identifier and relationship name too.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 2]
```diff
@@ -1389,152 +1389,201 @@ public Identifier getRelatedIdentifier(Contentlet contentlet,String relationship
...
-        return q;
-    }
+        for (SearchHit sh : response.getHits()) {
+            try {
+                Map<String, Object> sourceMap = sh.getSourceAsMap();
+                if (sourceMap.get("identifier") != null) {
+                    result.add(findContentletByIdentifierAnyLanguage(sourceMap.get("identifier").toString()));
+                }
+            } catch (Exception e) {
+                Logger.error(this, e.getMessage(), e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Please log the content's Identifier and relationship name too.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 3]
```diff
@@ -1389,152 +1389,201 @@ public Identifier getRelatedIdentifier(Contentlet contentlet,String relationship
...
-        String sortBy = rel.getRelationTypeValue() + "-" + contentlet.getIdentifier() + "-order";
-
         try {
-            List<ContentletSearch> contentletSearchList =
-                    searchIndex(q, -1, 0, sortBy, user, respectFrontendRoles);
-            return contentletSearchList.stream().map(ESContentletAPIImpl::transformContentletSearchToContent)
-                    .collect(CollectionsUtils.toImmutableList());
+            return getRelatedContentESQuery(contentlet, rel, user, respectFrontendRoles, pullByParent);
         } catch (Exception e){
             final String errorMessage = "Unable to look up related content for contentlet with identifier "
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Please log the relationship name too.

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/content/util/ContentUtils.java [hunk 4]
```diff
@@ -475,53 +476,50 @@ public static long count(String query, User user, String tmDate) {
...
-	}
 
-	/**
+            if (UtilMethods.isSet(condition)) {
+                pullQuery.append(" AND ").append(condition);
+            }
+
+            return pull(pullQuery.toString(), offset, limit, sort, user, tmDate);
+        } catch (Exception e) {
+            Logger.error(ContentUtils.class, "Error pulling related content", e);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/content/util/ContentUtils.java]: Please log the content's Identifier and relationship name too.

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/content/util/ContentUtils.java (hunk 4 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/content/util/ContentUtils.java]: Done

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 3 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Done

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 2 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: The catch was not necessary, so I removed it

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 1 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: The catch was not necessary, so I removed it

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 5]
```diff
@@ -2663,35 +2663,57 @@ public void unarchive(List<Contentlet> contentlets, User user,boolean respectFro
...
-    public void deleteRelatedContent(final Contentlet contentlet, final Relationship relationship, final boolean hasParent, final User user, final boolean respectFrontendRoles) throws DotDataException, DotSecurityException,DotContentletStateException {
-        if(!permissionAPI.doesUserHavePermission(contentlet, PermissionAPI.PERMISSION_EDIT, user, respectFrontendRoles)){
-            throw new DotSecurityException("User: " + (user != null ? user.getUserId() : "Unknown") + " cannot edit Contentlet");
+    public void deleteRelatedContent(final Contentlet contentlet, final Relationship relationship,
+            final boolean hasParent, final User user, final boolean respectFrontendRoles, final List<Contentlet> contentletsToBeRelated)
+            throws DotDataException, DotSecurityException, DotContentletStateException {
+        if (!permissionAPI.doesUserHavePermission(contentlet, PermissionAPI.PERMISSION_EDIT, user,
+                respectFrontendRoles)) {
+            throw new DotSecurityException("User: " + (user != null ? user.getUserId() : "Unknown")
+                    + " cannot edit Contentlet");
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Please add the Identifier of the Contentlet, for troubleshooting purposes.

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

**Review by dsilvam — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**